### PR TITLE
assists: Promote module to folder

### PR DESCRIPTION
**Pull request by longfangsong**
Close part of #10143.

This PR adds a assist to promote module to directory, which means make a .rs file module into a directory style module with the same name.

未命名(1)

### CODE REVIEW

File: crates/ide_assists/src/handlers/promote_mod_file.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+pub(crate) fn promote_mod_file(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let source_file = ctx.find_node_at_offset::<ast::SourceFile>()?;
+    let module = ctx.sema.to_module_def(ctx.frange.file_id)?;
+    if module.is_mod_rs(ctx.db()) {
+        return None;
+    }
+    let target = TextRange::new(
+        source_file.syntax().text_range().start(),
+        source_file.syntax().text_range().end(),
+    );
```
Comment (bjorn3) [inline crates/ide_assists/src/handlers/promote_mod_file.rs:35]: I think making it triggerable for the whole file is way to noisy. Maybe make it triggerable on the first line or the `mod` item?

**Comment by flodiebold** [comment]
I'd call this something like "Turn `a.rs` into `a/mod.rs`" to make clearer what it does. It's IMO not a "promotion" in any sense, and a mostly unnecessary transformation nowadays.

**Comment by longfangsong** [comment]
> I'd call this something like "Turn `a.rs` into `a/mod.rs`" to make clearer what it does. It's IMO not a "promotion" in any sense

Intellij Rust and [emacs](https://internals.rust-lang.org/t/is-mod-rs-the-right-name-for-the-directory-owning-module/4871/11?u=longfangsong) uses the word "promote". Though I'm not sure whether it is a good name either, maybe we should keepace with them?

**Comment by flodiebold** [comment]
I assume both IntelliJ and Emacs had this function before the 2018 edition. At that time, you could view it as a 'promotion' since it wasn't possible to have submodules without it. Plus, I don't think reusing the terminology from them really helps anyone -- it's not like someone used to "promote" wouldn't understand "Turn `a.rs` into `a/mod.rs`".

**Comment by matklad** [comment]
I do think we need something like this, but I don't think it's a good idea to make it available anywhere in a module -- this adds a :bulb: entry everywhere, which is unhelpful every time you *don't* invoke this assist. 

Let's re-use @vsrs's brilliant insight in #9735 and make the assist avialable *only when the whole file is selected (<kbd>ctrl+a</kbd>)*. 

I've documented this and other tricks in https://github.com/matklad/rust-analyzer/commit/46cdde75f83a18a97c32dadebfdec0a54c3be7c1

**Comment by longfangsong** [comment]
Now this assist is avialable only when the whole file is selected.
/cc @bjorn3 @matklad

File: crates/ide_assists/src/handlers/promote_mod_file.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        let end = trimmed_range.end() - TextSize::from(1);
+        trimmed_range = TextRange::new(trimmed_range.start(), end);
+    }
+    trimmed_range
+}
+
+// Assist: promote_mod_file
+//
+// Moves inline module's contents to a separate file.
```
Comment (jonas-schievink) [inline crates/ide_assists/src/handlers/promote_mod_file.rs:40]: This should be updated, it was copied from `move_module_to_file`

File: crates/ide_assists/src/handlers/promote_mod_file.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    #[test]
+    fn cannot_promote_mod_rs() {
+        check_assist_not_applicable(
+            promote_mod_file,
+            r#"//- /main.rs
+mod a;
+//- /a/mod.rs
+$0fn t() {}
```
Comment (jonas-schievink) [inline crates/ide_assists/src/handlers/promote_mod_file.rs:131]: This is testing the wrong thing. Consider adding a `cov_mark` here.

```suggestion
$0fn t() {}$0
```

**Comment by jonas-schievink** [comment]
Thanks, looks great now!

bors r+

**Comment by lnicola** [comment]
changelog feature (first contribution) add assist to convert file module to directory

**Comment by lnicola** [comment]
mod-to-folder

**Pull request merged**